nixos/modules/services/matrix/draupnir.nix

```diff
@@ -0,0 +1,210 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+
+let
+  cfg = config.services.draupnir;
+
+  format = pkgs.formats.yaml { };
+  configFile = format.generate "draupnir.yaml" cfg.settings;
+
+  inherit (lib)
+    literalExpression
+    mkEnableOption
+    mkOption
+    mkPackageOption
+    mkRemovedOptionModule
+    mkRenamedOptionModule
+    types
+    ;
+in
+{
+  imports = [
+    # Removed options for those migrating from the Mjolnir module
+    (mkRenamedOptionModule
+      [ "services" "draupnir" "dataPath" ]
+      [ "services" "draupnir" "settings" "dataPath" ]
+    )
+    (mkRenamedOptionModule
+      [ "services" "draupnir" "homeserverUrl" ]
+      [ "services" "draupnir" "settings" "homeserverUrl" ]
+    )
+    (mkRenamedOptionModule
+      [ "services" "draupnir" "managementRoom" ]
+      [ "services" "draupnir" "settings" "managementRoom" ]
+    )
+    (mkRemovedOptionModule [ "services" "draupnir" "pantalaimon" ] ''
+      `services.draupnir.pantalaimon.*` has been removed because it depends on the deprecated and vulnerable
+      libolm library for end-to-end encryption and upstream support for Pantalaimon in Draupnir is limited.
+      See <https://the-draupnir-project.github.io/draupnir-documentation/bot/encryption> for details.
+      If you nontheless require E2EE via Pantalaimon, you can configure `services.pantalaimon-headless.instances`
+      yourself and use that with `services.draupnir.settings.pantalaimon` and `services.draupnir.pantalaimonPasswordFile`.
+    '')
+  ];
+
+  options.services.draupnir = {
+    enable = mkEnableOption "Draupnir, a moderations bot for Matrix";
+
+    package = mkPackageOption pkgs "draupnir" { };
+
+    accessTokenFile = mkOption {
+      type = types.nullOr types.path;
+      default = null;
+      description = ''
+        File containing the access token for Draupnir's Matrix account
+        to be used in place of {option}`services.draupnir.settings.accessToken`.
+      '';
+    };
+
+    pantalaimonPasswordFile = mkOption {
+      type = types.nullOr types.path;
+      default = null;
+      description = ''
+        File containing the password for Draupnir's Matrix account when used in
+        conjunction with Pantalaimon to be used in place of
+        {option}`services.draupnir.settings.pantalaimon.password`.
+
+        ::: {.warning}
+        Read upstream's note on Pantalaimon and E2EE support (or lack thereof) in general:
+        <https://the-draupnir-project.github.io/draupnir-documentation/bot/encryption> and
+        <https://the-draupnir-project.github.io/draupnir-documentation/shared/dogfood#e2ee-support>.
+        :::
+      '';
+    };
```

Comment:
I kept it at top-level.
Should be easier to spot this way and I don't think there are even remotely enough secrets to justify making this truly generic.
I am open to move those into `services.draupnir.secrets.accessToken` and `services.draupnir.secrets.pantalaimon.password` nonetheless.
Just not free-form.

https://search.nixos.org/options?channel=unstable&show=services.forgejo.secrets for reference by what I mean by "truly generic".

Reply:
With the addition of third secret for `--http-antispam-authorization-path`/`web.synapseHTTPAntispam.authorization` via https://github.com/NixOS/nixpkgs/pull/400194#issuecomment-2945336107 I decided to move the secrets from top-level to `services.draupnir.secrets`, matching the structure of `services.draupnir.settings`:

- `cfg.secrets.accessToken` for `cfg.settings.accessToken`
- `cfg.secrets.pantalaimon.password` for `cfg.settings.pantalaimon.password`
- `cfg.secrets.web.synapseHTTPAntispam.authorization` for `cfg.settings.web.synapseHTTPAntispam.authorization`